src/backend/services/users-api/tests/functional/test_routes.py

```diff
@@ -0,0 +1,260 @@
+import pytest
+from unittest.mock import patch
+from src.databases import database_init
+import mongomock
+import json
+
+with patch.object(database_init, 'create_users_db', return_value=mongomock.MongoClient()):
+    from src.app import app
+    from src.app import mongo
+
+import sys, os
+
+
+from shared import HttpCode, FailureReturnString
+
+if os.environ.get('SERVICE_IN_DOCKER',False):             
+    sys.path.append(os.path.abspath(os.path.join('..', '')))
+else:
+    sys.path.append(os.path.abspath(os.path.join('../..', '')))
+
+@pytest.fixture(scope='module')
+def test_client():
+    test_client = app.test_client()
+
+    ctx = app.app_context()
+    ctx.push()
+
+    yield test_client
+
+    ctx.pop()
+
+@pytest.fixture(scope='module')
+def test_db_patch():
+    mongo.db.create_collection("users")
+
+    mongo.db.users.insert_one({"first_name" : "Akshay", "last_name" : "Sharma", "umnetID" : "SHARMAA2", "password" : "akshay123", "public_key" : "akshay_pk"})
+    mongo.db.users.insert_one({"first_name" : "Abhi", "last_name" : "Sachdev", "umnetID" : "SACHDEV1", "password" : "abhi123", "public_key" : "abhi_pk"})
+    mongo.db.users.create_index("umnetID", unique=True)
+    mongo.db.users.create_index("public_key", unique=True)
+ 
+@pytest.fixture(scope='module')
+def json_header():
+
+    mimetype = 'application/json'
+    headers = {
+        'Content-Type': mimetype,
+        'Accept': mimetype
+    }
+
+    return headers    
+
+
+def test_home_page(test_client, test_db_patch):
+    # test GET query on '/' route
+
+    url = '/'
+
+    response = test_client.get(url)
+
+    assert response.status_code == 200
+    resp_msg = "Hello Users of " + mongo.db.name
+    assert resp_msg.encode('utf-8') in response.data
+
+
+def test_insert_user(test_client, test_db_patch, json_header):
+    url = '/create'
+    
+    payload = {
+        "first_name" : "Madison",
+        "last_name" : "Fines",
+        "umnetID" : "FINESM1",
+        "password" : "madison123",
+        "public_key" : "madison_pk"
+        }
+
+    response = test_client.post(url, data=json.dumps(payload), headers=json_header)
+    assert b'User Madison Fines added successfully! :)' in response.data
+
+    user = mongo.db.users.find_one({"umnetID" : "FINESM1"})
+    assert user is not None
+
+    # remove the added user to maintain the state of the db
+    mongo.db.users.delete_one({"umnetID" : "FINESM1"})
+
+
+def test_get_all_users(test_client, test_db_patch):
+    
+    url = '/all'
```

Comment:
Don't forget to update this and other stuff in tests if you enact my suggestions above

Reply:
done